From dd8fa4c75eec6ddc2675363787f3a40cda793996 Mon Sep 17 00:00:00 2001
From: Wei Wen Goh <goweiwen@gmail.com>
Date: Sun, 23 Jul 2023 01:04:29 +0800
Subject: [PATCH] feat: add pause, unpause commands

---
 command.c | 14 ++++++++++++++
 command.h |  5 +++++
 2 files changed, 19 insertions(+)

diff --git a/command.c b/command.c
index e9facea0412..248b190caa0 100644
--- a/command.c
+++ b/command.c
@@ -681,6 +681,22 @@ bool command_show_osd_msg(command_t *cmd, const char* arg)
     return true;
 }
 
+bool command_pause(command_t *cmd, const char* arg)
+{
+   runloop_state_t *runloop_st = runloop_state_get_ptr();
+   runloop_st->flags |= RUNLOOP_FLAG_PAUSED;
+   runloop_st->flags |= RUNLOOP_FLAG_IDLE;
+   return true;
+}
+
+bool command_unpause(command_t *cmd, const char* arg)
+{
+   runloop_state_t *runloop_st = runloop_state_get_ptr();
+   runloop_st->flags &= ~RUNLOOP_FLAG_PAUSED;
+   runloop_st->flags &= ~RUNLOOP_FLAG_IDLE;
+   return true;
+}
+
 bool command_get_info(command_t *cmd, const char *arg)
 {
    char reply[128]              = "";
diff --git a/command.h b/command.h
index 1c57a064f8e..88c0b87692d 100644
--- a/command.h
+++ b/command.h
@@ -408,6 +408,8 @@ bool command_version(command_t *cmd, const char* arg);
 bool command_get_status(command_t *cmd, const char* arg);
 bool command_get_config_param(command_t *cmd, const char* arg);
 bool command_show_osd_msg(command_t *cmd, const char* arg);
+bool command_pause(command_t *cmd, const char* arg);
+bool command_unpause(command_t *cmd, const char* arg);
 bool command_get_info(command_t *cmd, const char* arg);
 bool command_get_disk_count(command_t *cmd, const char *arg);
 bool command_get_disk_slot(command_t *cmd, const char *arg);
@@ -448,6 +450,9 @@ static const struct cmd_action_map action_map[] = {
    { "READ_CORE_MEMORY", command_read_memory,      "<address> <number of bytes>" },
    { "WRITE_CORE_MEMORY",command_write_memory,     "<address> <byte1> <byte2> ..." },
 
+   { "PAUSE",            command_pause,            "No argument" },
+   { "UNPAUSE",          command_unpause,          "No argument" },
+
    { "GET_INFO",         command_get_info,         "No argument" },
    { "GET_DISK_COUNT",   command_get_disk_count,   "No argument" },
    { "GET_DISK_SLOT",    command_get_disk_slot,    "No argument" },
